### ISSUE TITLE
Trying to do inference time optimization on top of LTX2

### ISSUE THREAD

**Issue by asafjo23** [comfyanonymous/ComfyUI#11911]
### Custom Node Testing

- [x] I have tried disabling custom nodes and the issue persists (see [how to disable custom nodes](https://docs.comfy.org/troubleshooting/custom-node-issues#step-1%3A-test-with-all-custom-nodes-disabled) if you need help)

### Your question

Hi there, I hoped to get some help with the following issue. I am trying to enable gradients while forwarding LTX2:

```py
class SimpleLayoutOptimizer:
    """Optimizes latents to align cross-attention with desired spatial layouts."""

    def __init__(
        self,
        loss_fn: LayoutLoss,
        num_refinements: int = 3,
        lr_start: float = 0.01,
        lr_end: float = 0.05,
        betas: tuple[float, float] = (0.4, 0.9),
        weight_decay: float = 0.0,
    ):
        self.loss_fn = loss_fn
        self.num_refinements = num_refinements
        self.lr_start = lr_start
        self.lr_end = lr_end
        self.betas = betas
        self.weight_decay = weight_decay

    @torch.inference_mode(False)
    def optimize(
        self,
        model_function,
        latents: torch.Tensor,
        store: AttentionStore,
        cfg: float,
    ) -> torch.Tensor:
        latents = latents.clone()

        optimizer = torch.optim.AdamW(
            [latents],
            lr=self.lr_start,
            betas=self.betas,
            weight_decay=self.weight_decay,
        )
        scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
            optimizer, T_max=self.num_refinements, eta_min=self.lr_end
        )

        first_loss = None
        pbar = comfy.utils.ProgressBar(self.num_refinements)

        with torch.enable_grad():
            latents_model_input = latents.requires_grad_(True)
            for i in range(self.num_refinements):
                optimizer.zero_grad()
                store.reset()

                _ = model_function(latents_model_input)

                attn = store.get_avg_attention()

                if cfg > 1:
                    attn = attn[1:2]

                loss = self.loss_fn(attn)
                loss.backward()

                if first_loss is None:
                    first_loss = loss.item()

                current_lr = optimizer.param_groups[0]["lr"]
                grad_norm = latents.grad.norm().item()

                logger.info(
                    f"Refine {i+1}/{self.num_refinements}: "
                    f"loss {first_loss:.4f}→{loss.item():.4f}, "
                    f"grad {grad_norm:.2e}, "
                    f"lr {current_lr:.2e}"
                )

                optimizer.step()
                scheduler.step()
                pbar.update(1)

        store.reset()
        return latents.detach()


class LayoutGuidedGuider(comfy.samplers.CFGGuider):
    """Wraps a base guider and adds layout optimization at specified steps."""

    def __init__(
        self,
        base_guider: comfy.samplers.CFGGuider,
        optimizer: SimpleLayoutOptimizer,
        store: AttentionStore,
        refine_steps: List[int],
        model,
        original_attn_fns: Dict[str, Any],
        save_attention_config: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(base_guider.model_patcher)

        if hasattr(base_guider, "original_conds"):
            self.original_conds = base_guider.original_conds
        if hasattr(base_guider, "conds"):
            self.conds = base_guider.conds
        if hasattr(base_guider, "cfg"):
            self.cfg = base_guider.cfg

        self.base_guider = base_guider
        self.optimizer = optimizer
        self.store = store
        self.refine_steps = set(refine_steps)
        self.current_step = 0
        self.save_attention_config = save_attention_config
        self.model = model
        self.original_attn_fns = original_attn_fns

        if save_attention_config:
            os.makedirs(save_attention_config["output_dir"], exist_ok=True)
            logger.info(f"Attention saving enabled: {save_attention_config['output_dir']}")

    def save_attention_for_step(self, step_index: int):
        if not self.save_attention_config or self.store.accumulator is None:
            return

        config = self.save_attention_config
        attn = self.store.get_avg_attention()

        if attn.dim() == 4:
            attn = attn.mean(dim=1)

        if self.cfg > 1:
            attn = attn[1:2]
            logger.debug(f"CFG enabled, using conditional attention. Shape: {attn.shape}")

        for phrase, guiding_ind in zip(config["phrases"], config["token_indices"]):
            if not guiding_ind:
                logger.warning(f"No token indices found for phrase '{phrase}', skipping")
                continue

            lat_f, lat_h, lat_w = config["latent_shape"]
            attn_h = lat_h // 2
            attn_w = lat_w // 2
            token_attn = attn[:, :, guiding_ind].mean(-1).reshape(lat_f, attn_h, attn_w, 1)
            att_pil = attn_to_pil(token_attn)
            output_path = os.path.join(config["output_dir"], f"{phrase}_{step_index:03d}.mp4")
            save_attention_video(att_pil, output_path, fps=config["fps"])
            logger.debug(f"Saved attention: {output_path}")

        self.store.reset()

    def predict_noise(self, x, timestep, model_options={}, seed=None):
        if self.current_step in self.refine_steps:
            logger.info(f"Running layout optimization at step {self.current_step}")

            @torch.inference_mode(False)
            def model_fn(latents):
                logger.info(f"Latents is_inference: {latents.is_inference()}, requires_grad: {latents.requires_grad}")
                return super(LayoutGuidedGuider, self).predict_noise(
                    latents, timestep, model_options, seed
                )

            try:
                x = self.optimizer.optimize(
                    model_function=model_fn,
                    latents=x,
                    store=self.store,
                    cfg=self.cfg,
                )
            except Exception as e:
                logger.error(f"Optimization failed at step {self.current_step}: {e}")

        if self.save_attention_config:
            self.store.reset()

        result = super().predict_noise(x, timestep, model_options, seed)

        if self.save_attention_config:
            self.save_attention_for_step(self.current_step)

        self.current_step += 1

        return result

    def inner_sample(self, noise, latent_image, device, sampler, sigmas, denoise_mask, callback, disable_pbar, seed, latent_shapes=None):
        from .attention_store import unpatch_attention_layers

        self.current_step = 0
        try:
            return super().inner_sample(
                noise, latent_image, device, sampler, sigmas,
                denoise_mask, callback, disable_pbar, seed, latent_shapes
            )
        finally:
            if self.original_attn_fns:
                try:
                    unpatch_attention_layers(self.model, self.original_attn_fns)
                    logger.info(f"Unpatched {len(self.original_attn_fns)} attention layers")
                except Exception as e:
                    logger.error(f"Failed to unpatch attention layers: {e}")


class AttentionStore:
    """Accumulates attention probability maps across transformer layers."""

    def __init__(self):
        self.accumulator = None
        self.count = 0
        self.keep_heads = False

    def __call__(self, probs: torch.Tensor, layer_idx: Optional[int] = None) -> None:
        if not self.keep_heads:
            probs = probs.mean(dim=1)

        if self.accumulator is None:
            self.accumulator = probs
        else:
            self.accumulator = self.accumulator + probs
        self.count += 1

    def reset(self):
        self.accumulator = None
        self.count = 0

    def get_avg_attention(self) -> torch.Tensor:
        if self.accumulator is None or self.count == 0:
            raise ValueError("No attention maps stored")
        return self.accumulator / self.count


class AttentionCaptureWrapper:
    """Wraps an attention module to capture cross-attention probabilities."""

    def __init__(
        self,
        original_attn_fn,
        store: AttentionStore,
        layer_idx: int,
        capture_cross_attn_only: bool = True,
    ):
        self.original_attn_fn = original_attn_fn
        self.store = store
        self.layer_idx = layer_idx
        self.capture_cross_attn_only = capture_cross_attn_only

    def __call__(
        self,
        x: torch.Tensor,
        context: Optional[torch.Tensor] = None,
        mask: Optional[torch.Tensor] = None,
        **kwargs
    ) -> torch.Tensor:
        is_cross_attention = context is not None

        if not self.capture_cross_attn_only or is_cross_attention:
            self._capture_attention_probs(x, context, **kwargs)

        return self.original_attn_fn(x, context=context, mask=mask, **kwargs)

    def _capture_attention_probs(
        self,
        x: torch.Tensor,
        context: Optional[torch.Tensor],
        **kwargs
    ) -> None:
        attn_module = self.original_attn_fn.__self__

        q = attn_module.to_q(x)
        k = attn_module.to_k(x if context is None else context)

        if hasattr(attn_module, 'q_norm'):
            q = attn_module.q_norm(q)
        if hasattr(attn_module, 'k_norm'):
            k = attn_module.k_norm(k)

        pe = kwargs.get('pe')
        if pe is not None:
            from comfy.ldm.lightricks.model import apply_rotary_emb
            q = apply_rotary_emb(q, pe)
            k = apply_rotary_emb(k, kwargs.get('k_pe', pe))

        batch_size = q.shape[0]
        q = q.reshape(batch_size, -1, attn_module.heads, attn_module.dim_head).transpose(1, 2)
        k = k.reshape(batch_size, -1, attn_module.heads, attn_module.dim_head).transpose(1, 2)

        scale = 1.0 / math.sqrt(attn_module.dim_head)
        scores = torch.matmul(q, k.transpose(-1, -2)) * scale
        probs = F.softmax(scores, dim=-1)

        self.store(probs, self.layer_idx)


def patch_attention_layers(
    model,
    store: AttentionStore,
    layers_to_hook: list[int],
    cross_attention_only: bool = True,
) -> Dict[str, Any]:
    diffusion_model = _get_diffusion_model(model)
    original_attn_fns = {}

    from comfy.ldm.lightricks.model import CrossAttention

    layer_idx = 0
    for name, module in diffusion_model.named_modules():
        if not isinstance(module, CrossAttention):
            continue

        if cross_attention_only and not _is_video_text_attention(name):
            continue

        if layer_idx in layers_to_hook:
            original_attn_fns[name] = module.forward
            module.forward = AttentionCaptureWrapper(
                original_attn_fn=module.forward,
                store=store,
                layer_idx=layer_idx,
                capture_cross_attn_only=cross_attention_only,
            )

        layer_idx += 1

    return original_attn_fns


def unpatch_attention_layers(model, original_attn_fns: Dict[str, Any]):
    diffusion_model = _get_diffusion_model(model)

    for name, original_fn in original_attn_fns.items():
        module = diffusion_model
        for attr in name.split('.'):
            if not hasattr(module, attr):
                raise AttributeError(f"Module path '{name}' not found in diffusion model")
            module = getattr(module, attr)
        module.forward = original_fn


def _get_diffusion_model(model):
    if hasattr(model, 'model') and hasattr(model.model, 'diffusion_model'):
        return model.model.diffusion_model
    if hasattr(model, 'diffusion_model'):
        return model.diffusion_model
    raise ValueError("Cannot find diffusion model in provided model object")


def _is_video_text_attention(module_name: str) -> bool:
    return module_name.split('.')[-1] == 'attn2'

```
But I am always getting the following error:
```
Optimization failed at step 0: Inference tensors cannot be saved for backward. Please do not use Tensors created in inference mode in computation tracked by autograd. To work around this, you can make a clone to get a normal tensor and use it in autograd, or use `torch.no_grad()` instead of `torch.inference_mode()`.
```

Hoped someone can assist and sheds some light over the issue.
Thanks in advance,

### Logs

```powershell

```
